build/commands/lib/depotTools.js

```diff
@@ -0,0 +1,147 @@
+const fs = require('fs')
+const path = require('path')
+const Log = require('./logging')
+const util = require('./util')
+
+// This guard is used to ensure that we don't have a partial or broken checkout
+// of depot_tools.
+const depotToolsGuard = new ActionGuard(
+  path.join(
+    path.dirname(config.depotToolsDir),
+    `${path.basename(config.depotToolsDir)}_install.guard`
+  ),
+  () => {
+    if (fs.existsSync(config.depotToolsDir)) {
+      Log.status('detected incomplete depot_tools install. Cleaning up...')
+      fs.rmSync(config.depotToolsDir, { recursive: true })
+    }
+  }
+)
+
+// The .disable_auto_update file in the depot_tools directory, regardless of
+// its content, disables auto-updates. If a specific depot_tools reference is
+// enforced, we use this file for two purposes: 1) disable auto-updates; 2)
+// store the enforced reference. Note: The presence of this file also prevents
+// the auto-bootstrapping of CIPD (Chrome Infrastructure Package Deployment),
+// Python, and other dependencies.
+const disableAutoUpdateFilePath = path.join(
+  config.depotToolsDir,
+  '.disable_auto_update'
+)
+
+// Reads the enforced depot_tools ref from .disable_auto_update file. If the
+// file does not exist or the ref is not valid, returns null.
+function readEnforcedDepotToolsRef() {
+  if (!fs.existsSync(disableAutoUpdateFilePath)) {
+    return null
+  }
+  const ref = fs.readFileSync(disableAutoUpdateFilePath, 'utf8').trim()
+  const isRefValid = /^[a-f0-9]{40}$|^refs\/tags\/.*$|^origin\/.*$/.test(ref)
+  return isRefValid ? ref : null
+}
+
+// Writes the enforced depot_tools ref to .disable_auto_update file. If ref is a
+// falsy value, the file is removed.
+function writeEnforcedDepotToolsRef(ref) {
+  if (ref) {
+    fs.writeFileSync(disableAutoUpdateFilePath, ref)
+  } else if (fs.existsSync(disableAutoUpdateFilePath)) {
+    fs.unlinkSync(disableAutoUpdateFilePath)
+  }
+}
+
+function installDepotTools(options = config.defaultOptions) {
+  options.cwd = config.braveCoreDir
+
+  const enforcedDepotToolsRef = config.getProjectRef('depot_tools', null)
+
+  // If we're modifying depot_tools reference, it's important to ensure a clean
+  // checkout to re-bootstrap CIPD, Python and other dependencies.
+  if (
+    enforcedDepotToolsRef !== readEnforcedDepotToolsRef() &&
+    fs.existsSync(config.depotToolsDir)
+  ) {
+    Log.status(
+      'depot_tools ref is updated. Removing existing checkout to ' +
+        're-bootstrap all dependencies.'
+    )
+    fs.rmSync(config.depotToolsDir, { recursive: true })
+  }
+
+  if (!fs.existsSync(config.depotToolsDir) || depotToolsGuard.isDirty()) {
+    Log.progressScope('install depot_tools', () => {
+      depotToolsGuard.ensureClean(() => {
+        util.run(
+          'git',
+          ['clone', config.depotToolsRepo, config.depotToolsDir],
+          options
+        )
+      })
+    })
+  }
+
+  if (enforcedDepotToolsRef !== readEnforcedDepotToolsRef()) {
```

Comment:
Could we move that code inside `depotToolsGuard` to be sure that `.disable_auto_update` is always in the consistent state?
The same with `ninjaLogCfgPath`.